### PR TITLE
add UniversalPageStorage

### PR DESCRIPTION
**Pull request by lidezhu**
### What problem does this PR solve?

Issue Number: ref #6728 

Problem Summary:

### What is changed and how it works?
1. Rename DB::PageId to PageIdU64;
2. Make many classes like `PageDirectory` `BlobStore` a template;
3. Add `UniversalPageStorage` which receive a string type page id;

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Storages/Page/V2/PageStorage.h [hunk 1]
```diff
@@ -85,14 +85,23 @@ class PageStorage : public DB::PageStorage
...
     };
 
+public:
+    // Use a more easy gc config for v2 when all of its data will be transformed to v3.
+    static PageStorageConfig getEasyGCConfig()
+    {
+        PageStorageConfig gc_config;
+        gc_config.file_roll_size = PAGE_FILE_SMALL_SIZE;
+        return gc_config;
+    }
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Page/V2/PageStorage.h:95]: Move it to be `PageStorageConfig::getEasyGCConfig()`

File: dbms/src/Storages/DeltaMerge/tests/gtest_segment_test_basic.cpp [hunk 2]
```diff
@@ -171,7 +171,7 @@ std::optional<PageId> SegmentTestBasic::splitSegmentAt(PageId segment_id, Int64
     return right->segmentId();
 }
 
-void SegmentTestBasic::mergeSegment(const std::vector<PageId> & segments_id, bool check_rows)
+void SegmentTestBasic::mergeSegment(const std::vector<PageIdU64> & segments_id, bool check_rows)
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/tests/gtest_segment_test_basic.cpp:174]:
```suggestion
void SegmentTestBasic::mergeSegment(const PageIdsU64 & segments_id, bool check_rows)
```

File: dbms/src/Storages/DeltaMerge/tests/gtest_segment_test_basic.h [hunk 3]
```diff
@@ -52,58 +52,58 @@ class SegmentTestBasic : public DB::base::TiFlashStorageTestBasic
...
+    bool isSegmentDefinitelyEmpty(PageIdU64 segment_id);
 
-    PageId getRandomSegmentId();
+    PageIdU64 getRandomSegmentId();
 
     /**
      * You must pass at least 2 segments. Checks whether all segments passed in are sharing the same stable.
      */
-    [[nodiscard]] bool areSegmentsSharingStable(const std::vector<PageId> & segments_id) const;
+    [[nodiscard]] bool areSegmentsSharingStable(const std::vector<PageIdU64> & segments_id) const;
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/tests/gtest_segment_test_basic.h:89]:
```suggestion
    [[nodiscard]] bool areSegmentsSharingStable(const PageIdU64s & segments_id) const;
```

File: dbms/src/Storages/Page/V2/PageStorage.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (lidezhu) [reply dbms/src/Storages/Page/V2/PageStorage.h]: Fixed.

File: dbms/src/Storages/Page/Page.h [hunk 4]
```diff
@@ -52,30 +52,36 @@ struct FieldOffsetInsidePage
...
+        page.page_id = INVALID_PAGE_U64_ID;
+        page.is_valid = false;
+        return page;
     }
 
     Page()
-        : page_id(INVALID_PAGE_ID)
+        : page_id(INVALID_PAGE_U64_ID)
+        , is_valid(true)
     {}
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Page/Page.h:66]: Why we need a valid Page with `INVALID_PAGE_U64_ID`?

Comment (lidezhu) [reply dbms/src/Storages/Page/Page.h:66]: OK, I will pass the page_id through the constructor.

File: dbms/src/Storages/Page/V3/GCDefines.cpp [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+        assert(callbacks.prefix != 0);
+    }
+    else if constexpr (std::is_same_v<Trait, universal::ExternalPageCallbacksManagerTrait>)
+    {
+        assert(callbacks.prefix != "");
+    }
+    // NamespaceId(TableID) should not be reuse
+    RUNTIME_CHECK_MSG(
+        callbacks_container.count(callbacks.prefix) == 0,
+        "Try to create callbacks for duplicated namespace id {}",
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Page/V3/GCDefines.cpp:99]:
```suggestion
        "Try to create callbacks for duplicated prefix {}",
```

File: dbms/src/Storages/Page/Page.h (hunk 4 above)
Comment (lidezhu) [reply dbms/src/Storages/Page/Page.h]: Fixed.

**Review by JaySon-Huang — APPROVED** [review]
LGTM

**Review by breezewish — APPROVED** [review]

File: dbms/src/Storages/Page/V3/PageDirectory.cpp [hunk 6]
```diff
@@ -1475,20 +1539,38 @@ bool PageDirectory::tryDumpSnapshot(const ReadLimiterPtr & read_limiter, const W
...
     // persist new logs into disk. So we pass `nullptr` as `wal` to the factory.
-    PageDirectoryPtr collapsed_dir = factory.createFromReader(
-        identifier,
-        std::move(snapshot_reader),
-        /* wal */ nullptr);
+    auto collapsed_dir = [&]() {
+        // we just use the `collapsed_dir` to dump edit of the snapshot, should never call functions like `apply` that
+        // persist new logs into disk. So we pass `nullptr` as `wal` to the factory.
+        //        static_assert(std::is_same_v<Trait, u128::PageDirectoryTrait> || std::is_same_v<Trait, universal::PageDirectoryTrait>,
+        //                      "unknown impl");
```
[the remark this answers is not in the document]
Comment (lidezhu) [reply dbms/src/Storages/Page/V3/PageDirectory.cpp]: Fixed.

File: dbms/src/Storages/Page/V3/Universal/UniversalPageId.h [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+namespace DB
+{
+class UniversalPageId final
+{
+public:
+    static inline UniversalPageId toFullPageId(const String & prefix, PageIdU64 page_id)
+    {
+        WriteBufferFromOwnString buff;
+        writeString(prefix, buff);
+        UniversalPageIdFormat::encodeUInt64(page_id, buff);
```
[the remark this answers is not in the document]
Comment (lidezhu) [reply dbms/src/Storages/Page/V3/Universal/UniversalPageId.h:32]: We will also provide page-ctl tools for uni ps to dump its data. So this should be ok.

File: dbms/src/Storages/Page/Page.h [hunk 8]
```diff
@@ -52,30 +52,35 @@ struct FieldOffsetInsidePage
...
+        Page page{INVALID_PAGE_U64_ID};
+        page.is_valid = false;
+        return page;
     }
 
-    Page()
-        : page_id(INVALID_PAGE_ID)
+    explicit Page(PageIdU64 page_id_)
+        : page_id(page_id_)
+        , is_valid(true)
```
[the remark this answers is not in the document]
Comment (lidezhu) [reply dbms/src/Storages/Page/Page.h:64]: The `Page` struct is both used in `PS V3` and `Uni PS`, so the PageId can be UInt128 or String. 
And we extract a `UInt64` from them and pass it to `Page` for easy adaption to previous code. So `page_id` should be a `UInt64` here. Or it should be removed totally.

**Pull request merged**